tests/e2e/framework/ipam/pool.go

```diff
@@ -85,45 +83,44 @@ func (f *pool) Release(ipnet string) error {
 	return nil
 }
 
-func (f *pool) randomIPv4(subnet *net.IPNet) (string, error) {
+func (f *pool) getIPv4(subnet *net.IPNet) (string, error) {
 	f.lock.Lock()
 	defer f.lock.Unlock()
 
 	if !containsSubnet(f.cidr, subnet) {
 		return "", fmt.Errorf("subnet %s not in ip pool %s", subnet, f.cidr)
 	}
 
-	// todo: replace retry with check has available IP first
-	for i := 0; i < 10; i++ {
-		ipv4 := randomIPv4FromSubnet(subnet)
-		ipNet := (&net.IPNet{IP: ipv4, Mask: f.cidr.Mask}).String()
-
+	ones, bits := subnet.Mask.Size()
+	poolSize := 1 << (bits - ones)
+	if len(f.ipUsed) >= poolSize {
```

Comment:
`f.ipUsed` is IPs used in f.cidr, not in `subnet`.

Reply:
remove full check at the beginning, failed after checking whole subnet